### DISCUSSION TITLE
PorousFlow： Porosity setting

### DISCUSSION THREAD

**Discussion by YNianer** [idaholab/moose#25228] [Q&A Modules: General]
Hello, how to set the porosity related to stress?
<img width="551" alt="微信图片_20230817174052" src="https://github.com/idaholab/moose/assets/53603198/7019b2bf-567e-4c0e-8187-842cd437d535">

**Comment by josebastiase** [comment]
Hi,
I don't think this function is implemented in MOOSE. You have two options:
1.- You can code a new material object (with a bit of C++)
2.- If Residual and zero porosity and constant, you can use a `ParsedFunction` and pass the result to a AuxKernel and AuxVariable and use `PorousFlowPorosityConst`.
And probably a second `ParsedFunction` to compute the mean stress.

**Reply by GiudGiud** [reply]
If you intend to solve for stress and porosity at the same time AND you are using AD, option 2 (with the functions) will not naturally propagate derivatives .
You will need to code the contributions to the Jacobian manually. 

It seems there are porosity as a function of strain in the module, but not stress. @cpgr any suggestions on this?

**Reply by cpgr** [reply]
If you need that relationship you will have to code a new porosity material yourself.
